kwasm-lemmas.md

```diff
@@ -31,6 +44,42 @@ In this case, it's simpler (and safe) to simply discard the `#chop`, instead of
     rule #minSigned(ITYPE) => 0 -Int #pow1(ITYPE)
 ```
 
+Memory
+------
+
+Memory is represented by a byte map, where each key is an index and each entry either empty (0) or a byte (1 to 255).
+This invariant must be maintained by the semantics, and any failure to maintain it is a breach of the Wasm specification.
+We want to make the variant explicit, so we introduce the following helper, which simply signifies that we assume a well-formed byte map:
+
+```k
+    syntax Bool ::= "#isWellFormedByteMap" "(" ByteMap ")" [function, smtlib(isWellFormedByteMap)]
+ // ----------------------------------------------------------------------------------------------
```

Comment:
Do you want to add rules about what a well-formed `ByteMap` looks like? Something like:

```
rule #isByteMap(ByteMap <| .Map |>) => true
rule #isByteMap(ByteMap <| _ |-> V M |>) => #isByte(V) andBool #isByteMap(M)
rule #isByteMap(ByteMap <| M [ _ <- V ] |>) => #isByte(V) andBool #isByteMap(M)
```

Note that I also am suggesting the shorter name `#isByteMap` over `#isWellFormedByteMap`.

Reply:
Good idea. I was mostly afraid this would trip up the prover (that it would start applying these rules and get stuck, rather than just take it at face value), but the proofs seem to go through still.